File: 321. Create Maximum Number/Java/Solution.java
public class Solution {
    public int[] maxNumber(int[] nums1, int[] nums2, int k) {
        class Utils {
            int[] maxNumberLengthK(int[] nums, int k) {
                int[] result = new int[k];
                int index = 0;
                int length = nums.length;
                for (int i = 0; i < length; i++) {
                    while (index > 0 && result[index - 1] < nums[i] && length - i >= k - (index - 1)) {
                        index--;
                    }
                    if (index < k) {
                        result[index] = nums[i];
                        index++;
                    }
                }
                return result;
            }

            int[] merge(int[] nums1, int[] nums2) {
                int[] result = new int[nums1.length + nums2.length];
                int i = 0;
                int j = 0;
                int index = 0;
                while (i < nums1.length && j < nums2.length) {
                    if (greater(nums1, i, nums2, j)) {
                        result[index] = nums1[i];
                        i++;
                        index++;
                    } else {
                        result[index] = nums2[j];
                        j++;
                        index++;
                    }
                }
                if (i == nums1.length) {
                    while (j < nums2.length) {
                        result[index] = nums2[j];
                        j++;
                        index++;
                    }
                } else {
                    while (i < nums1.length) {
                        result[index] = nums1[i];
                        i++;
                        index++;
                    }
                }
                return result;
            }

            boolean greater(int[] nums1, int start1, int[] nums2, int start2) {
                while (start1 < nums1.length && start2 < nums2.length && nums1[start1] == nums2[start2]) {
                    start1++;
                    start2++;
                }
                if (start1 == nums1.length && start2 == nums2.length) {
                    return true;
                }
                if (start1 == nums1.length) {
                    return false;
                }
                if (start2 == nums2.length) {
                    return true;
                }
                return nums1[start1] > nums2[start2];
            }
        }
        Utils utils = new Utils();
        int[] result = new int[k];
        for (int i = 0; i <= k; i++) {
            if (i <= nums1.length && k - i <= nums2.length) {
                int[] current = utils.merge(utils.maxNumberLengthK(nums1, i), utils.maxNumberLengthK(nums2, k - i));
                if (utils.greater(current, 0, result, 0)) {
                    result = current;
                }
            }
        }
        return result;
    }
}

File: 321. Create Maximum Number/Python/Solution.py
class Solution(object):
    def maxNumber(self, nums1, nums2, k):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :type k: int
        :rtype: List[int]
        """
        def max_number_length_k(nums, k):
            result = [0] * k
            index = 0
            length = len(nums)
            for i, num in enumerate(nums):
                while index > 0 and result[index - 1] < num and length - i >= k - (index - 1):
                    index -= 1
                if index < k:
                    result[index] = num
                    index += 1
            return result

        def merge(nums_1, nums_2):
            result = []
            i = 0
            j = 0
            while i < len(nums_1) and j < len(nums_2):
                if greater(nums_1, i, nums_2, j):
                    result.append(nums_1[i])
                    i += 1
                else:
                    result.append(nums_2[j])
                    j += 1
            if i == len(nums_1):
                result.extend(nums_2[j:])
            else:
                result.extend(nums_1[i:])
            return result

        def greater(nums_1, start_1, nums_2, start_2):
            while start_1 < len(nums_1) and start_2 < len(nums_2) and nums_1[start_1] == nums_2[start_2]:
                start_1 += 1
                start_2 += 1
            if start_1 == len(nums_1) and start_2 == len(nums_2):
                return True
            if start_1 == len(nums_1):
                return False
            if start_2 == len(nums_2):
                return True
            return nums_1[start_1] > nums_2[start_2]

        result = [0] * k
        for i in range(k + 1):
            if i <= len(nums1) and k - i <= len(nums2):
                current = merge(max_number_length_k(nums1, i), max_number_length_k(nums2, k - i))
                if greater(current, 0, result, 0):
                    result = current
        return result
